epee: Change Color Contrast On Magenta (Console Color)

Tzadiko opened the pull request:
**Description**
I use the monero-wallet-cli, and I have trouble reading a lot of the information, simply because the colors do not have enough contrast. For example, you can see here that the magenta blends in with the background. 

![image](https://github.com/user-attachments/assets/dd71bd77-d0dd-4a72-a411-365d1196b66f)

Even my non-colorblind friend has a hard time reading this.

![image](https://github.com/user-attachments/assets/fce92545-75f6-4fca-8991-ffa5655c0916)

I looked at the API and noticed that there are better colors that can be used. For example, all these color values would be preferred over the current color scheme. 

![image](https://github.com/user-attachments/assets/99af3cce-5521-4a54-bf7a-a57da3d8bbc4)

Have a look at the **spent** transation.

### Before
![image](https://github.com/user-attachments/assets/0c1d1a2a-9c23-42f5-8a5b-9b07286ab009)

### After

**Windows**
![image](https://github.com/user-attachments/assets/3f4eaf77-a5a2-4438-9313-e7a10dc0d46a)

**Linux**
![image](https://github.com/user-attachments/assets/ef7a9b49-a628-464a-9c8f-ea5b5487a179)

Tzadiko committed:
Change the magenta color to a lighter version of itself.

This helps those that are colorblind or cannot tell the difference
between the shell background and the current color.

selsta commented:
Would it be easier to just change the low contrast colors instead of adding a flag?

Tzadiko commented:
> Would it be easier to just change the low contrast colors instead of adding a flag?

Im open to it, I just don't know how others would feel about changing the existing colors, so I figured I'd fit a 'high contrast' color mode in. I'd like more people to chime in before I commit to continuing this change. Some more spots where I want to integrate the IMessageWriter.

Tzadiko commented:
> As a color blind user I can say I would most likely not remember to or be too lazy to type the flag relative to just struggling through. Therefore, I vote to just change the low contrast colors.
> […](#)
> On Mar 13, 2025, at 4:33 PM, Tzadiko ***@***.***> wrote: Tzadiko left a comment ([monero-project/monero#9831](https://github.com/monero-project/monero/pull/9831)) Would it be easier to just change the low contrast colors instead of adding a flag? Im open to it, I just don't know how others would feel about changing the existing colors, so I figured I'd fit a 'high contrast' color mode in. I'd like more people to chime in before I commit to continuing this change. Some more spots where I want to integrate the IMessageWriter. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you are subscribed to this thread. <[#9831 (comment)](https://github.com/monero-project/monero/pull/9831#issuecomment-2722922664)> <https://github.com/notifications/unsubscribe-auth/AAAUIMIT5GD32JEKGSGCKET2UIITZAVCNFSM6AAAAABY5FF3JOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRSHEZDENRWGQ> Tzadiko left a comment ([monero-project/monero#9831](https://github.com/monero-project/monero/pull/9831)) <[#9831 (comment)](https://github.com/monero-project/monero/pull/9831#issuecomment-2722922664)> Would it be easier to just change the low contrast colors instead of adding a flag? Im open to it, I just don't know how others would feel about changing the existing colors, so I figured I'd fit a 'high contrast' color mode in. I'd like more people to chime in before I commit to continuing this change. Some more spots where I want to integrate the IMessageWriter. — Reply to this email directly, view it on GitHub <[#9831 (comment)](https://github.com/monero-project/monero/pull/9831#issuecomment-2722922664)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUIMIT5GD32JEKGSGCKET2UIITZAVCNFSM6AAAAABY5FF3JOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRSHEZDENRWGQ>. You are receiving this because you are subscribed to this thread.

That works. Alternatively I can just set --colorblind to `true` by default. I kind of like the code I wrote so far, haha. I think it makes what's going on clearer and opens the code to extension down the line, if someone wants to implement a different `IMessageWriter`. One thing I didn't like that I changed with this (that wouldn't be changed if I just replaced magenta) is [**this**](https://github.com/monero-project/monero/pull/9831/files#diff-25b9648b3da507060201d4ce30539c281676672c1af9a8e4d0e16ec0585a987bL8709).

Tzadiko commented:
@selsta @monerobby This MR is no longer a WIP. Please see my reply as to why I believe this change is preferred to just simply changing a color code ([here](https://github.com/monero-project/monero/pull/9831#issuecomment-2723362211)).

I changed --colorblind to --highcontrast and made is `true` by default.

There are some free functions and macros where I cannot retrieve the --highcontrast argument (as that is a property of simplewallet.h). Those methods call failure_msg_writer and message_writer (methods that existed prior) whose bodies not call `MessageWriterFactory` to retrieve the high-contrast writer. Placing calls to MessageWriterFactory in these existing methods reduces the changeset (the alternative is to replace each of these methods with calls to MessageWriterFactory).

iamamyth commented:
> I think it makes what's going on clearer and opens the code to extension down the line, if someone wants to implement a different IMessageWriter (e.g. instead of writing to console, write to a database, or file, etc). In other words, it adds value beyond just changing a color.

I would say that this is a false generalization, and therefore subtracts value: The only consumer of this "interface" is a console application. It does not want to write to a database. Its output may, in fact, end up on the filesystem, if stdout or stderr has been redirected, but the OS does the work in that case. Similarly, if a user really wanted to redirect the output from the CLI to "anywhere other than the console", that's why operating system output redirection exists. Simply put: Console applications should write to stdout/stderr, as appropriate. Do not add complexity without utility.

Tzadiko commented:
> Do not add complexity without utility.

Okay I'll either amend this MR or open a separate one to just amend the colors, without the IMessageWriter interface and additional code that comes with it.

selsta commented on contrib/epee/src/mlog.cpp:
```diff
@@ -441,12 +441,13 @@ void set_console_color(int color, bool bright)
       HANDLE h_stdout = GetStdHandle(STD_OUTPUT_HANDLE);
-      SetConsoleTextAttribute(h_stdout, FOREGROUND_BLUE | FOREGROUND_RED | (bright ? FOREGROUND_INTENSITY:0));
+      int colorblindMagentaCode = 13;
+      SetConsoleTextAttribute(h_stdout, colorblindMagentaCode | (bright ? FOREGROUND_INTENSITY:0));
```
`13` is `0x000D`, which equals `FOREGROUND_RED | FOREGROUND_BLUE | FOREGROUND_INTENSITY`. Because the intensity bit is already set, the conditional OR has no effect and both values of bright produce the same bright-magenta attribute.

selsta commented on contrib/epee/src/mlog.cpp:
```diff
@@ -441,12 +441,13 @@ void set_console_color(int color, bool bright)
 #else
       if(bright)
-        std::cout << "\033[1;35m";
+        std::cout << "\033[38;2;245;176;236m";
```
This appears to introduce the first use of 24-bit ANSI color in the codebase. Is this safe without checking for truecolor?
